client/src/index.ts

```diff
@@ -90,19 +90,78 @@ type FilterConstraints<T> = {
   [key in keyof T]?: T[key] extends Record<string, any> ? FilterConstraints<T[key]> : T[key] | DeepConstraint<T[key]>;
 };
 
+type CursorNavigationOptions = { first?: string } | { last?: string } | { after?: string; before?: string };
+type OffsetNavigationOptions = { size?: number; offset?: number };
+type PaginationOptions = CursorNavigationOptions & OffsetNavigationOptions;
+
 type BulkQueryOptions<T> = {
+  page?: PaginationOptions;
+  /** TODO: Not implemented yet
   filter?: FilterConstraints<T>;
   sort?:
     | {
         column: keyof T;
         direction?: SortDirection;
       }
     | keyof T;
+**/
 };
 
 type QueryOrConstraint<T, R> = Query<T, R> | Constraint<T>;
 
-export class Query<T, R = T> {
+type QueryMeta = { page: { cursor: string; more: boolean } };
+
+interface BasePage<T, R> {
+  query: Query<T, R>;
+  meta: QueryMeta;
+  records: R[];
+
+  nextPage(options?: OffsetNavigationOptions): Promise<Page<T, R>>;
```

Comment:
When we use cursor-based pagination we don't want to pass an `offset`, I think. Just a `size`.

I think we only should allow `offset` in the `getMany` method. All xxxPage methods should only receive an optional `offset`. WDYT?

Reply:
I thought the same originally, BUT the API actually allows sending an ``offset`` with cursor based pagination. So I ended-up including them both in the method signature to be more flexible. In any case we could spread the parameters so it's easier ``nextPage(size?: number, offset?: number);``